### DISCUSSION TITLE
Moving  Nitro to standalone bundler

### DISCUSSION THREAD

**Discussion by pi0** [nuxt/nuxt#18219] [RFCs]
Almost one year ago (Nov 2, 2020) we initiated a project called `Sigma` with the goal of making it possible to deploy Nuxt 2 SSR apps to Cloudflare workers. This idea evolved so fast and we finally renamed it to `Nitro` as the new Server stack for Nuxt 3. The initial version of nitro was a standalone nuxt module it then moved it to Nuxt 3 monorepo allowing faster iterations and easier testing. 

Today, Nitro exports a semi-standalone API and auto-published as `@nuxt/nitro-edge` and bundling Nuxt's SSR app using rollup, h3, unstorage, and unenv with platform-specific runtime and also adds the possibility to add server API and Middleware.

## **Current problems**

Currently, we are hard-coding Nuxt specific SSR logic directly in [Nitro SSR Runtime](https://github.com/nuxt/framework/blob/main/packages/nitro/src/runtime/app/render.ts) and many vue-specific hacks into [Nitro Rollup Config](https://github.com/nuxt/framework/blob/main/packages/nitro/src/rollup/config.ts) and it is slowly making code less maintainable over time and more likely for regressions. 

We also have a test matrix that tests (Vite/Webpack/Bridge-Vite/Bridge-Webpack) against at least 4 targets (16 runs). This amount of CI time is unnecessary if nitro could be tested in an isolated environment against targets and only tested against Nuxt once per setup making CI 4x faster and allowing adding more presets without more slow down.

## **New possibilities**

By this separation, we can implement new ideas such as WASM support (#692), Selective SSR for server-only components (#666), Functions (nuxt/rfcs#35), Using Vite, and Nitro plugins and then easily use in nuxt without risk of regressions and less trouble of depending on a whole SSR stack.

Also by refactoring SSR logic out and introducing plugins like `nitro-plugin-vue` we can decouple server-side-rendering and a potentially open door to supporting new frameworks like React (same path of vite)

This also means nitro can be used to **deploy any server** to **any platform** which brings a chance to a much wider amount of use cases that can bring back more love and features to Nuxt 3! And at the same time, unifying efforts of platforms to help maintain a single place without bothering them with nuxt specific requirements.

For Nuxt, we have fewer issues, faster CI, and less code to maintain while we can focus on providing the best DX experience.

For short-term maintenance, since the nitro core is almost stable and tested over months against different targets, we can start `0.x` releases with proper changelog and stop using edge builds.

In terms of long-term maintenance, while it sounds like extra effort, it would certainly solve many versioning problems since once nitro is stable, nuxt integration is as easy as using nitro as a dependency just like webpack and vite bundlers. Changelog and breaking changes being separated can also allow us easily adapt with new major versions of nitro in the same semver-major release of `nuxt3`.

## **Migration plan**

Nitro [context](https://github.com/nuxt/framework/blob/2166661b1781a98dcc736a8b3262704d18b89655/packages/nitro/src/context.ts) is decoupled from nuxt internals. So the initial refactoring should be almost straight forward and we only need to decouple hard-coded vue/nuxt logic as a nitro plugin and set up new repository for it and progressively iterate to make new base cleaner.
For deployment documentation, we shall keep it in nuxt until finding a better place.

## **How it will look like:**

For name, I could only find [nitropack](https://npmjs.com/package/nitropack) available on npm. Still open to better name suggestions :)

The new package will expose a programmatic API and a mini CLI `nitropack build` and `nitropack dev`. The core of nitro pack will provide an essential setup including h3 and storage and SSR integration will be added as a Nitro plugin (`nitro-plugin-vue`). For nuxt integration, we import core API, add vue plugin and a nuxt plugin to hook into the lifecycle, and add nuxt logics like payload extraction and meta integration.

**Comment by atinux** [comment]
Closing since it is now on https://github.com/unjs/nitro
